### PR TITLE
fix utf8 -> utf16 decoding bug on surrogate pairs

### PR DESCRIPTION
**Pull request by turbio**
This fixes https://github.com/protobufjs/protobuf.js/issues/1473

The custom utf8 -> utf16 decoder appears to be subtly flawed. From my reading it appears the chunking mechanism doesn't account for surrogate pairs at the end of a chunk causing variable size chunks. A larger chunk followed by a smaller chunk leaves behind garbage that'll be included in the latter chunk.

The chunking mechanism was added to prevent stack overflows when calling `formCharCode` with too many args. From some simple benchmarking it doesn't seem like putting utf16 code units in an array and spreading that into `fromCharCode` is helping performance much anyway. I've removed the chunking code which simplifies it significantly and fixes the particular issue we've encountered.

Here's a repro of the existing encoding bug in a fuzzing suite
https://repl.it/@turbio/oh-no-our-strings#decoder.js

### CODE REVIEW

File: lib/utf8/index.js [hunk 1]
```diff
@@ -38,36 +38,27 @@ utf8.length = function utf8_length(string) {
...
-        chunk = [],
-        i = 0, // char offset
-        t;     // temporary
-    while (start < end) {
-        t = buffer[start++];
-        if (t < 128)
-            chunk[i++] = t;
-        else if (t > 191 && t < 224)
-            chunk[i++] = (t & 31) << 6 | buffer[start++] & 63;
-        else if (t > 239 && t < 365) {
```
Comment (turbio) [inline lib/utf8/index.js:54]: My understanding of this function is that the passed in buffer is always an array of uint8s, each byte being a utf8 code unit. The condition `t < 365` is a bit odd under this assumption, from profiling it actually forces v8 to deopt when `t` is compared to a non uint8 value. Alternatively I may have grossly misunderstood how this is expected to work.

**Comment by alexander-fenster** [comment]
Hi @turbio,

Would it be possible to have a test that shows the problem included in the PR?

Thank you!

**Comment by turbio** [comment]
@alexander-fenster yup!
 
added a test which fails on the old `utf8_read` function.

**Review by alexander-fenster — APPROVED** [review]
Thanks @turbio!

**Comment by jparser** [comment]
hi guys, the merged code seems not included in the latest release.
protobufjs@7.4.0 -> @protobufjs/utf8@1.1.0,  @protobufjs/utf8@1.1.0  is still the old code.

image
image

**Pull request merged**